### PR TITLE
User login and nickname feature with Authentication API

### PR DESCRIPTION
**Pull request by ChenyuHeidiZhang**
Screenshot 2020-06-09 at 9 42 44 PM - Display 1

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 1]
```diff
@@ -66,16 +68,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    UserService userService = UserServiceFactory.getUserService();
 
+    // Only logged-in users can post comments.
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/comments.html");
+      return;
+    }
+
+    String userId = userService.getCurrentUser().getUserId();
+    String mood = request.getParameter("mood");  // Get the input parameters from the form.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Move the comment to above line 79 since it applies to more than just this single statement.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns user's login status and corresponding Urls for log in or out. */
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]:
```suggestion
/** Servlet that returns user's login status and corresponding URLs for logging in or out. */
```

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String urlToRedirectToAfterUserLogsOut = "/comments.html";
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: You could shorten the name to `redirectUrlLogout`, the meaning is still clear

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      response.getWriter().println("{\"isLogin\": true, " + 
+          "\"logoutUrl\": \"" + logoutUrl + "\", " + 
+          "\"userId\": " + "\"" + userService.getCurrentUser().getUserId() + "\"}");
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/comments.html";
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Similarly, 'redirectUrlLogin'

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    String userId = request.getParameter("userId");    
+    String nickname = getUserNickname(userId);
+    String displayName;
+    // Display nickname if it is set; display email otherwise.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: Just a nit:
(A bit cleaner in my opinion)

```suggestion
    // Display nickname if set, email otherwise.
```

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that stores the nickname input by a user alongside the user's id and returns a display name upon GET request. */
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]:
```suggestion
/** Servlet that stores the user's id and nickname entered, and returns a display name upon GET request. */
```

Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: It might also be better to move the "returns ..." part to be a JSdoc for the `doGet` function.

@lawrencezjh thoughts?

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      return;
+    }
+
+    String nickname = request.getParameter("nickname-input");
+    String userId = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", userId);
+    entity.setProperty("userId", userId);
+    entity.setProperty("nickname", nickname);
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: Define "userId" and "nickname" as String constants at the top of the file since you're using them multiple times.

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    // The put() function automatically inserts new data or updates existing data based on IDs.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: As this comment stands, it makes it seem like you're defining what the `put` does.
Instead, you want to talk about what you're doing.

I would say something along the lines of:
```
Insert a new entry for the user in the data store.
Note: The "userId" is used as a key to identify each entry, so existing entries will be updated to use the new "nickname".
```

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    datastore.put(entity);
+
+    response.sendRedirect("/comments.html");
+  }
+
+  /**
+   * Returns the nickname of the user with id, or empty String if the user has not set a nickname.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]:
```suggestion
   * Returns the nickname of the user with the specified id, or empty String no nickname is set.
```

Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: You also want to write this entirely as an `@return`.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -250,3 +251,40 @@ function deleteComment(comment) {
...
+      loginOutLink.href = loginStatus.loginUrl;
+      loginOutLink.innerText = 'Login';
+    }
+  });
+}
+
+/** 
+ * Checks log in status and fetches all the comments when comments.html is on load.
+ */
+function commentsPageOnLoad() {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: Since you're checking login status and fetching comments, a name like `fetchCommentsIfLoggedIn` would be more appropriate.

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -250,3 +251,40 @@ function deleteComment(comment) {
...
+      document.getElementById('input-form-fieldset').setAttribute('disabled', 'true');
+      document.getElementById('set-nickname-button').style.display = 'none';
+      loginOutLink.href = loginStatus.loginUrl;
+      loginOutLink.innerText = 'Login';
+    }
+  });
+}
+
+/** 
+ * Checks log in status and fetches all the comments when comments.html is on load.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: "when comments.html is on load" is a bit unclear. I would rephrase this.

File: portfolio/src/main/webapp/comments.html [hunk 12]
```diff
@@ -56,26 +56,53 @@
...
+            <div class="col-sm-8">
+              <h4 id="welcome-message">Welcome, </h4>
+            </div>
+            <div class="col-sm-4">          
+              <a id="login-out" href="#" class="btn btn-secondary">Login</a>
+              <button id="set-nickname-button" class="btn btn-light" data-toggle="modal" data-target="#nickname-modal">Set Nickname</button>
+            </div>
+          </div>
+
+          <!-- The modal for setting nickname -->
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/comments.html]:
```suggestion
          <!-- The modal for setting nickname. -->
```

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -250,3 +251,40 @@ function deleteComment(comment) {
   params.append('id', comment.id);
   fetch('/delete-comments', {method: 'POST', body: params});
 }
+
+/**
+ * Checks if the user is logged in and displays corresponding contents.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: What are these "contents" that are being displayed?

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -250,3 +251,40 @@ function deleteComment(comment) {
   params.append('id', comment.id);
   fetch('/delete-comments', {method: 'POST', body: params});
 }
+
+/**
+ * Checks if the user is logged in and displays corresponding contents. 
+ */
+function checkLoginStatus() {
+  fetch('/login').then(response => response.json()).then(loginStatus => {
+    const isLogin = loginStatus.isLogin;
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: Rename to: `isLoggedIn`

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -250,3 +251,40 @@ function deleteComment(comment) {
   params.append('id', comment.id);
   fetch('/delete-comments', {method: 'POST', body: params});
 }
+
+/**
+ * Checks if the user is logged in and displays corresponding contents. 
+ */
+function checkLoginStatus() {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: If you're also displaying messages after checking the status, you might want to rename this.

`checkLoginStatus` would mean that all you do is check a true/false of whether the user is logged in. Nothing further.

File: portfolio/src/main/webapp/style.css [hunk 16]
```diff
@@ -252,6 +252,14 @@ h3 {
 
 /* comments.html */
 
+#comments-header {
+  margin-top:20px;
+}
+
+#login-out, #set-nickname-button {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/style.css]: Each id should be on a new line.
```suggestion
#login-out,
#set-nickname-button {
```

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 1 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Got it

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 2 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: okie

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Yup, the current name is too long lol. I'm adding a comment to be more clear.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 4 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 5 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: Sounds good!

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 6 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: You mean Javadoc? I was wondering if we could put comments on override methods like doGet and doPost, since none of the examples do that (but I don't see why not).

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 7 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: Got it

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 9 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: I think "Returns" is the right way. From the Javadoc style guide:
Tip: A common mistake is to write simple Javadoc in the form /** @return the customer ID */. This is incorrect, and should be changed to /** Returns the customer ID. */.

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Comments would be fetched even if user is not logged in. If there has to be a name it would be `checkLoginAndFetchComments`. But since the only reason that this method exists is because the two methods should be both called when the page is onload, I think `commentsPageOnLoad` is fine.

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: when comments.html is loading?

File: portfolio/src/main/webapp/comments.html (hunk 12 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/comments.html]: Got it

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Added more clarification

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Got it

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Changed it to `setupCommentsPage`

File: portfolio/src/main/webapp/script.js (hunk 15 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: changed it to setupPageByLoginStatus()

File: portfolio/src/main/webapp/style.css (hunk 16 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/style.css]: ok

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String redirectUrlLogout = "/comments.html";  // The URL to redirect to after the user logs out.
+      String logoutUrl = userService.createLogoutURL(redirectUrlLogout);
+      response.getWriter().println("{\"isLoggedIn\": true, " + 
+          "\"logoutUrl\": \"" + logoutUrl + "\", " + 
+          "\"userId\": " + "\"" + userService.getCurrentUser().getUserId() + "\"}");
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: can you:
1) extract this string to be a private constant string 
2) use String.format

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    } else {
+      String redirectUrlLogin = "/comments.html";  // The URL to redirect to after the user logs in.
+      String loginUrl = userService.createLoginURL(redirectUrlLogin);
+      response.getWriter().println("{\"isLoggedIn\": false, \"loginUrl\": \"" + loginUrl + "\"}");
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: same here for this string

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+  private String getUserNickname(String userId) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate(USERID, Query.FilterOperator.EQUAL, userId));
+
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return "";
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: don't return an empty string and compare it against "".
Instead, just return null, and mark this API as @Nullable.

File: portfolio/src/main/webapp/comments.html [hunk 20]
```diff
@@ -13,7 +13,7 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
   </head>
-  <body onload="commentsPageOnLoad()">
+  <body onload="setupCommentsPage()">
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/comments.html]: Nit: How about "loadCommentsPage"? 
I say this because you're actually filling up the content of the page, and not just setting up (with content loading elsewhere).

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 17 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Yes that's a good idea

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 18 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: okie

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 19 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: okk

File: portfolio/src/main/webapp/comments.html (hunk 20 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/comments.html]: Sounds good

**Comment by ChenyuHeidiZhang** [comment]
Just fixed a bug that I found. That was the longest commit message I've ever typed 0.0

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -250,3 +251,41 @@ function deleteComment(comment) {
   params.append('id', comment.id);
   fetch('/delete-comments', {method: 'POST', body: params});
 }
+
+/**
+ * Checks if the user is logged in and displays corresponding welcome messages and login/logout buttons.
+ * Also, if user is logged in, enables posting comments and displays set-nickname button.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]:
```suggestion
 * Enables posting comments and displays set-nickname button if the user is logged in.
```

File: portfolio/src/main/webapp/script.js [hunk 22]
```diff
@@ -250,3 +251,41 @@ function deleteComment(comment) {
...
+  });
+}
+
+/** 
+ * Checks log in status and fetches all the comments to set up comments.html when the page is loading.
+ */
+function loadCommentsPage() {
+  setupPageByLoginStatus()
+  fetchComments(true);
+}
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: Just to confirm, does this also actually display the comments? 

If so, you want to update the JS doc to say that part.

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 21 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: ok

File: portfolio/src/main/webapp/script.js (hunk 22 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Ok

**Review by lawrencezjh — APPROVED** [review]

File: portfolio/src/main/webapp/script.js [hunk 23]
```diff
@@ -283,7 +283,8 @@ function setupPageByLoginStatus() {
 }
 
 /** 
- * Checks log in status and fetches all the comments to set up comments.html when the page is loading.
+ * Checks log in status to set up comments.html when the page is loading; 
+ * fetches all the comments and displays them.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]:
```suggestion
 * Checks log in status to set up comments.html when the page is loading. 
 * Fetches all the comments and displays them.
```

**Review by hayleejane3 — CHANGES_REQUESTED** [review]
Make the change one change and it should be good to go :)

File: portfolio/src/main/webapp/script.js (hunk 23 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: Done

File: portfolio/src/main/webapp/script.js [hunk 24]
```diff
@@ -292,11 +293,21 @@ function setupPageByLoginStatus() {
...
+      });
+}
+
 /** 
- * Checks log in status to set up comments.html when the page is loading; 
- * fetches all the comments and displays them.
+ * Checks log in status to set up comments.html when the page is loading. 
+ * Fetches all the comments and displays them. Fetches the blobstore URL.
  */
 function loadCommentsPage() {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: I think you want to rename this now since you're doing so much more in there.

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 24 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: It's hard to think of a good name lol. I call it `initiateCommentsPage` now.

File: portfolio/src/main/webapp/comments.html [hunk 25]
```diff
@@ -13,7 +13,7 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
   </head>
-  <body onload="fetchComments(true); fetchBlobstoreUrl();">
+  <body onload="loadCommentsPage();">
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/comments.html]: I think you need to rename this to initiateCommentsPage() as well?

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/comments.html (hunk 25 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/comments.html]: omg I must have forgot

**Review by hayleejane3 — APPROVED** [review]
